### PR TITLE
Transaction builder ESDT issue

### PR DESCRIPTION
**Pull request by mariusmihaic**
- Added `gasLimit` field in `TxBuilderInput`, which is by default = `DEFAULT_GAS_LIMIT`. If gas limit input from factory !=  `DEFAULT_GAS_LIMIT`, tx builder will use that one, otherwise, it will compute it using the `gasEstimator`
- Added esdt issue payload builder
- Added tx builder

### CODE REVIEW

File: src/transaction/payload_builder.cpp [hunk 1]
```diff
@@ -1,9 +1,33 @@
...
+
+std::string ESDTPropertyField(std::string const &property, bool const &val)
+{
+    std::string const value = val ? "true" : "false";
+    return "@" + util::stringToHex(property) + "@" + util::stringToHex(value);
+}
+
+std::string ESDTPropertiesAsOnData(ESDTProperties const &esdtProperties)
+{
+    return ESDTPropertyField("canFreeze", esdtProperties.canFreeze) +
```
Comment (dragos-rebegea) [inline src/transaction/payload_builder.cpp:19]: add canTransferNFTCreateRole

File: src/transaction/payload_builder.cpp [hunk 2]
```diff
@@ -103,3 +127,47 @@ std::string MultiESDTNFTTransferPayloadBuilder::build() const
...
+std::string ESDTIssuePayloadBuilder::build() const
+{
+    SCArguments args;
+    args.add(m_token);
+    args.add(m_ticker);
+    args.add(m_initialSupply);
+    args.add(BigUInt(m_numOfDecimals));
+
+    std::string data = ESDT_ISSUANCE_PREFIX + args.asOnData();
+    data += (m_esdtProperties != ESDT_ISSUANCE_DEFAULT_PROPERTIES) ? ESDTPropertiesAsOnData(m_esdtProperties) : "";
```
Comment (dragos-rebegea) [inline src/transaction/payload_builder.cpp:171]:
```suggestion
    if (m_esdtProperties != ESDT_ISSUANCE_DEFAULT_PROPERTIES) {
        data += ESDTPropertiesAsOnData(m_esdtProperties)
    }
```

**Review by dragos-rebegea — CHANGES_REQUESTED** [review]

File: src/transaction/payload_builder.cpp (hunk 1 above)
Comment (mariusmihaic) [reply src/transaction/payload_builder.cpp:19]: Right, was not aware of it. Thanks, good suggestion :+1:

File: src/transaction/payload_builder.cpp (hunk 2 above)
Comment (mariusmihaic) [reply src/transaction/payload_builder.cpp:171]: Matter of preference, I would keep it in 1 line as it is

Comment (dragos-rebegea) [reply src/transaction/payload_builder.cpp:171]: :+1:

**Review by dragos-rebegea — APPROVED** [review]

File: include/transaction/payload_builder.h [hunk 3]
```diff
@@ -61,4 +62,26 @@ class MultiESDTNFTTransferPayloadBuilder
     ContractCall m_contractCall;
 };
 
+class ESDTIssuePayloadBuilder
```
Comment (andreibancioiu) [inline include/transaction/payload_builder.h:65]: :rocket:

File: src/transaction/transaction_builders.cpp [hunk 4]
```diff
@@ -12,14 +12,17 @@ std::shared_ptr<bytes> stringToBytesPtr(std::string const &in)
...
         ITransactionBuilder(),
         m_txInput(std::move(txInput))
 {}
 
 Transaction TransactionEGLDTransferBuilder::build()
 {
-    uint64_t gasLimit = m_txInput.gasEstimator.forEGLDTransfer(m_txInput.data.size());
+    uint64_t inputGasLimit = m_txInput.gasLimit;
+    uint64_t estimatedGasLimit = m_txInput.gasEstimator.forEGLDTransfer(m_txInput.data.size());
+    uint64_t gasLimit = (inputGasLimit == DEFAULT_GAS_LIMIT) ? estimatedGasLimit : inputGasLimit;
```
Comment (andreibancioiu) [inline src/transaction/transaction_builders.cpp:25]: Somehow, if the user provides a gas limit equal to DEFAULT_GAS_LIMIT, that's ignored. However, the result is still good :)

So, tricky, but correct - can you double check, though?

**Review by andreibancioiu — APPROVED** [review]
:+1:

File: src/transaction/transaction_builders.cpp (hunk 4 above)
Comment (mariusmihaic) [reply src/transaction/transaction_builders.cpp:25]: User cannot provide gasLimit, only gasPrice. GasLimit is always set by the factory in builders

**Pull request merged**